README.md

# flowSpy <img src="https://github.com/JhuangLab/flowSpy/blob/master/inst/figures/logo.png" align="right" height=150 width=150/>

flowSpy is an R package to implement cellular subpopulations identification, trajectory inference, pseudotime estimation and visualization for flow and mass cytometry data. This package is developed and maintained by [JhuangLab](https://github.com/JhuangLab) at Shanghai Institute of Hematology.

Instructions and standard workflow can be found at:

https://github.com/JhuangLab/flowSpy/tree/master/vignettes

See the quick start tutorial of flowSpy, please visit [Quick start of flowSpy](https://ytdai.github.io/flowSpy/Quick_start.html).

See the basic tutorial of flowSpy, please visit [Tutorial of flowSpy](https://ytdai.github.io/flowSpy/basic.html).

See time-course data analysis of flowSpy, please visit [Time-course workflow of flowSpy](https://ytdai.github.io/flowSpy/Time_course.html).


Use cases could be found at: 

https://github.com/JhuangLab/flowSpy-dataset


You can view and clone the repository of flowSpy on GitHub at:

https://github.com/JhuangLab/flowSpy


## Note

----------------

Dear flowSpy users:

To improve the identification of this package and avoid awkward duplication of names in some situations, we decided to change the name of `flowSpy` to `CytoTree`. The package name of `CytoTree` more fits the functional orientation of this software. The usage and update of `flowSpy` and `CytoTree` will be consistent until the end of Bioc 3.11. And for the 3.12 devel, flowSpy will be deprecated.

The package `CytoTree` has been added to Bioconductor (https://bioconductor.org/packages/CytoTree/), we recommend that users can download this package and replace `flowSpy` as soon as possible.

We apologized for the inconvenience.

flowSpy team

2020-07-09

----------------


## 1 Introduction

Multidimensional single-cell-based flow and mass cytometry  enable ones to analyze multiple single-cell parameters and identify cellular populations. 
Based on classical software for analyzing [Flow Cytometry Standard](https://en.wikipedia.org/wiki/Flow_Cytometry_Standard) (FCS) data such as [`flowSOM`](https://bioconductor.org/packages/release/bioc/html/FlowSOM.html)[1] and [`SPADE`](https://github.com/nolanlab/spade)[2], methods for inferencing cellular trajectory during a biological process are very important. 
To objectively inference differential trajectory based on time courses FCS data, we present [`flowSpy`](https://github.com/JhuangLab/flowSpy), a trajectory inference and visualization toolkit for flow and mass cytometry data. 

`flowSpy` can help you to perform four main types of analysis:

- **Clustering**. `flowSpy` can help you to discover and identify subtypes of cells. 

- **Dimensionality Reduction**. Several dimensionality reduction methods are provided in `flowSpy` package such as Principal Components Analysis (PCA), t-distributed Stochastic Neighbor Embedding (tSNE), Diffusion Maps and Uniform Manifold Approximation and Projection (UMAP). flowSpy provides both cell-based and cluster-based dimensionality reduction.

- **Trajectory Inference**. `flowSpy` can help you to construct the cellular differential based on minimum spanning tree (MST) algorithm. 

- **Pseudotime and Intermediate states definition**. The root cells need to be defined by users. The trajctroy value will be calculated based on Shortest Path from root cells and leaf cells using R `igraph` package. Subset FCS data set in `flowSpy` and find the key intermediate cell states based on trajectory value.

## 2 Installation

### 2.1 From Bioconductor

```

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("flowSpy")

```

### 2.2 From Github

This requires the `devtools` package to be installed first.

```

# If not already installed
install.packages("devtools") 
devtools::install_github("JhuangLab/flowSpy")

library(flowSpy)

```


## 3 Quick start (Standard Workflow)

``` {r}

# Loading packages
suppressMessages({
library(ggplot2)
library(flowSpy)
library(flowCore)
library(stringr)
})

# Read fcs files
fcs.path <- system.file("extdata", package = "flowSpy")
fcs.files <- list.files(fcs.path, pattern = '.FCS$', full = TRUE)

fcs.data <- runExprsMerge(fcs.files, comp = FALSE, transformMethod = "none")

# Refine colnames of fcs data
recol <- c(`FITC-A<CD43>` = "CD43", `APC-A<CD34>` = "CD34", 
           `BV421-A<CD90>` = "CD90", `BV510-A<CD45RA>` = "CD45RA", 
           `BV605-A<CD31>` = "CD31", `BV650-A<CD49f>` = "CD49f",
           `BV 735-A<CD73>` = "CD73", `BV786-A<CD45>` = "CD45", 
           `PE-A<FLK1>` = "FLK1", `PE-Cy7-A<CD38>` = "CD38")
colnames(fcs.data)[match(names(recol), colnames(fcs.data))] = recol
fcs.data <- fcs.data[, recol]

day.list <- c("D0", "D2", "D4", "D6", "D8", "D10")
meta.data <- data.frame(cell = rownames(fcs.data),
                        stage = str_replace(rownames(fcs.data), regex(".FCS.+"), "") )
meta.data$stage <- factor(as.character(meta.data$stage), levels = day.list)

markers <- c("CD43","CD34","CD90","CD45RA","CD31","CD49f","CD73","CD45","FLK1","CD38")

# Build the FSPY object
fspy <- createFSPY(raw.data = fcs.data, markers = markers,
                   meta.data = meta.data,
                   normalization.method = "log",
                   verbose = TRUE)

# See information
fspy

# Standard workflow of flowSpy
fspy <- runCluster(fspy)
fspy <- processingCluster(fspy)
fspy <- runFastPCA(fspy)
fspy <- runTSNE(fspy)
fspy <- runDiffusionMap(fspy)
fspy <- runUMAP(fspy)
fspy <- buildTree(fspy, dim.type = "umap", dim.use = 1:2)
fspy <- defRootCells(fspy, root.cells = "Root cells")
fspy <- runPseudotime(fspy)
fspy <- defLeafCells(fspy, leaf.cells = "Leaf cells")
fspy <- runWalk(fspy)


```

## 4 Reported bugs and solutions

If there is any error in installing or librarying the `flowSpy` package, please contact us via e-mail forlynna@sjtu.edu.cn

## 5 Version History

Jul 09, 2020
 - Version 1.2.3
 - Changes:
   - Add note and change package name to `CytoTree`

May 07, 2020
 - Version 1.2.2
 - Changes:
   - Add solutions for large sample size analysis

May 04, 2020
  - Version 1.2.1
  - Changes:
    - Set linear transformation of FSC and SSC as optional in `runExprsMerge` and `runExprsExtract`
    - Fixed bugs in vignette

Apr 26, 2020
  - Version 1.1.5
  - Changes:
    - Add preprocessing tutorial
    - Fixed bugs in `runExprsExtract` and `runExprsMerge`

Apr 20, 2020
  - Version 1.1.5
  - Changes:
    - Fixed installation errors in Linux and Windows
    - Fixed legend name of figure
    - Fixed errors in building vignette

Mar 19, 2020
  - Version 1.0.4
  - Changes:
    - Fixed installation errors in Linux and Windows

Mar 17, 2020
  - Version 1.0.3
  - Changes:
    - Fixed bugs in `runDiff`

Feb 15, 2020
  - Version 1.0.1
  - Changes:
    - Add uniform downsampling in function `processingCluster`

Oct 17, 2019
  - Version 0.99.3
  - Changes:
    - Fixed some warnings in the checking process of Bioc 3.10 

Oct 17, 2019
  - Version 0.99.2
  - Changes:
    - Added documentation for some parameters
    - Changed the default parameter to `raw` in `dim.type`, and add match.arg() to `dim.type` selection

Sep 19, 2019
  - Version 0.99.1
  - Changes:
    - Fixed some bugs

Sep 19, 2019
  - Version 0.2.8
  - Changes:
    - Removed flowSpy.Rproj, .DS_Store and .gitattributes

Sep 11, 2019
  - Version 0.2.7
  - Changes:
    - Fixed some bugs
    - Fixed errors and warnings in `R CMD check` and `R CMD BiocCheck`

Aug 29, 2019
  - Version 0.2.6
  - Changes:
    - Fixed some bugs
    - Add trajectory plot function
    - Update vignette tutorial

Aug 14, 2019
  - Version 0.2.5
  - Changes:
    - Fixed some bugs
    - Add branch analysis and differentially expressed markers analysis

Aug 08, 2019
  - Version 0.2.4
  - Changes:
    - Fixed some bugs and finished vignette tutorial

July 24, 2019
  - Version 0.2.3
  - Changes:
    - Add phenoGraph algorithm

July 19, 2019
  - Version 0.2.2
  - Changes:
    - Fixed some bugs on cluster based downsampling


## 6 Reference

[1] Sofie Van Gassen, Britt Callebaut and Yvan Saeys (2019). FlowSOM: Using
  self-organizing maps for visualization and interpretation of cytometry data.
  http://www.r-project.org, http://dambi.ugent.be.

[2] Qiu, P., et al., Extracting a cellular hierarchy from high-dimensional cytometry data with SPADE. Nat Biotechnol, 2011. 29(10): p.886-91.







NEWS
Changes in version 1.2.3 (2020-07-09)
+ Add note and change package name to `CytoTree`

Changes in version 1.2.2 (2020-05-07)
+ Add solutions for large sample size analysis

Changes in version 1.2.1 (2020-05-04)
+ Set linear transformation of FSC and SSC as optional in `runExprsMerge` and `runExprsExtract`
+ Fixed bugs in vignette

Changes in version 1.1.6 (2020-04-26)
+ Add preprocessing tutorial
+ Fixed bugs in `runExprsExtract` and `runExprsMerge`

Changes in version 1.1.5 (2020-04-20)
+ Fixed installation errors in Linux and Windows
+ Fixed legend name of figure
+ Fixed errors in building vignette

Changes in version 1.0.4 (2020-03-19)
+ Fixed installation errors in Linux and Windows

Changes in version 1.0.3 (2020-03-17)
+ Fixed bugs in `runDiff`

Changes in version 1.0.1 (2020-02-15)
+ Add uniform downsampling in function `processingCluster`

Changes in version 1.0.1 (2019-10-17)
+ Fixed some warnings in the checking process of Bioc 3.10 



R/hello.R
##############################
# This is a test panel
##############################
if (F) {

  rm(list = ls())
  library(roxygen2)
  roxygenize()


verbose = T
cell.number = 500

sample.list <- paste0("D", c(0, 2, 4, 6, 8, 10))
raw <- NULL
for (i in 1:length(sample.list)) {
  sub <- read.table(paste0("../flowSpy-dataset/CleanData/usecase2/", sample.list[i], ".sub", cell.number, ".txt"), header = T, stringsAsFactors = F)
  sub$sample <- sample.list[i]
  raw <- rbind(raw, sub)
}
raw$sample <- factor(as.character(raw$sample), levels = sample.list)
table(raw$sample)

markers <- c("CD34", "CD43", "CD38", "CD90", "CD49f", "CD31", "CD45RA", "FLK1", "CD73")
length(markers)

raw.data <- as.matrix(raw[, 1:(dim(raw)[2]-1) ])
rownames(raw.data) <- paste0(raw$sample, "_", 1:length(raw$sample))
meta.data <- data.frame(cell = paste0(raw$sample, "_", 1:length(raw$sample)),
                        stage = raw$sample)

fspy.meta.data <- meta.data
fspy.raw.data <- raw.data
#save(fspy.meta.data, fspy.raw.data, file = "data/FSPYdata.rda")
batch <- factor(fspy.meta.data$stage, labels = 1:length(unique(raw$sample)))



###################################
library(ggplot2)
#library(flowSpy)
#data("FSPYdata")

markers <- c("CD34", "CD43", "CD90", "CD38", "CD49f", "CD31", "CD45RA", "FLK1", "CD73")


object <- createFSPY(raw.data = fspy.raw.data, markers = markers,
                     meta.data = fspy.meta.data,
                     normalization.method = "log",
                     verbose = T)

set.seed(1)
object <- runCluster(object, cluster.method = "som", xdim = 6, ydim = 6, verbose = T)
table(object@meta.data$cluster.id)
set.seed(2)
object <- processingCluster(object, downsampleing.size = 0.5, seed = 2)

object

object <- runKNN(object, knn = 200, verbose = T)

object <- runFastPCA(object, verbose = T)

object <- runTSNE(object, verbose = T)

object <- runDiffusionMap(object, verbose = T)

object <- runUMAP(object, verbose = T)

object <- buildTree(object, dim.type = "umap", dim.use = 1:2, verbose = T)

##########
plot2D(object, item.use = c("UMAP_1", "UMAP_2"), color.by = "stage", alpha = 1, main = "PCA", category = "categorical", show.cluser.id = T)

plot2D(object, item.use = c("UMAP_1", "UMAP_2"), color.by = "som.id", alpha = 1, main = "PCA", category = "categorical", show.cluser.id = T)

# pseudotime

object <- defRootCells(object, root.cells = c(13), verbose = T)

object <- runPseudotime(object, verbose = T)

object <- defLeafCells(object, leaf.cells = c(32,26,27), verbose = T)

object <- runWalk(object, verbose = T)

fetch.cells <- fetchCell(object, traj.value.log = 0.1, is.root.cells = 1, is.leaf.cells = 1)
sub.obj <- subsetFSPY(object, cells = fetch.cells)

# 4,29,9,19

plotMarkerDensity(object)


plot2D(object, item.use = c("UMAP_1", "UMAP_2"), color.by = "pseudotime", alpha = 1, main = "PCA", category = "numeric", show.cluser.id = T)

#D10.percent.stage
plotTree(object, color.by = "D6.percent", show.node.name = T, cex.size = 1.5) + scale_colour_gradientn(colors = c("#00599F", "#EEEEEE", "#FF3222", "#FF3222"))

plotPieTree(object, cex.size = 1, size.by.cell.number = F) + scale_fill_manual(values = c("#00599F", "#33CC33", "#FF3222", "#F4D31D", "#FF3222","#7a06a0"))

plotPieCluster(object, item.use = c("PC_1", "PC_2"), cex.size = 0.5)
plotCluster(object, item.use = c("PC_1", "PC_2"), size = 10)


plotTree(object, color.by = "pseudotime", show.node.name = T, cex.size = 1, as.tree = T) + scale_colour_gradientn(colors = c("#F4D31D", "#FF3222","#7a06a0"))



plotPseudotimeDensity(object)
plotPseudotimeTraj(object, var.cols = T) + scale_colour_gradientn(colors = c("#F4D31D", "#FF3222","#7a06a0"))
plotPseudotimeTraj(object, cutoff = 0.1, var.cols = T) + scale_colour_gradientn(colors = c("#F4D31D", "#FF3222","#7a06a0"))

plot2D(object, item.use = c("pseudotime", "CD43"), color.by = "som.id", alpha = 1, main = "PCA", category = "categorical", show.cluser.id = T)

plot2D(object, item.use = c("CD43", "CD34"), color.by = "stage", alpha = 1, main = "PCA")

plot3D(object, item.use = c("PC1", "PC2", "PC3"), color.by = "CD43")

plot2D(object, item.use = c("tSNE_1", "tSNE_2"), color.by = "pseudotime", alpha = 1, main = "PCA") + scale_colour_gradientn(colors = c("blue", "blue", "white", "red"))

plot2D(object, item.use = c("pseudotime", "traj.value.log"), color.by = "stage")

plotTree(object, show.node.name = T, cex.size = 2) + scale_colour_gradientn(colors = "#666666")

plotTree(object, color.by = "pseudotime", as.tree = T, show.node.name = T)  + scale_colour_gradientn(colors = c("#00599F", "#00599F","#EEEEEE", "#FF3222","#FF3222"))

plot.info <- fetchPlotMeta(object, verbose = F)
ggplot(plot.info, aes(x=traj.value.log, colour = stage)) + geom_density() + theme_bw()

pdata <- aggregate(plot.info[, c(markers, "pseudotime")], list(id = plot.info$som.id), mean)
pdata <- pdata[, -1]

pdata <- pdata[order(pdata$pseudotime), ]

plot.info.sub <- plot.info[, c(markers, "pseudotime")]
plot.info.sub <- plot.info.sub[order(plot.info.sub$pseudotime), ]
plot.info.sub <- plot.info.sub[, match(c("CD34", "CD43","CD31","CD45RA","CD38","CD49f","CD90","FLK1","CD73"), colnames(plot.info.sub))]
pheatmap(t(plot.info.sub),
         color = colorRampPalette(c(rep("#00599F",3), "#FFFFFF", rep("#FF3222",3)))(100),
         cluster_rows = F, cluster_cols = F, scale = "row", fontsize_col = 0.01)


plot.info.sub <- plot.info[, c("pseudotime", "traj.value.log")]
plot.info.sub <- plot.info.sub[plot.info.sub$traj.value.log > 0.2, ]
plot.info.sub <- plot.info.sub[order(plot.info.sub$pseudotime),]
plot(plot.info.sub$pseudotime, ylim = c(0,1), xlim = c(0,12000), col = "#AA00C2")

###############
sub.obj <- runKNN(sub.obj, knn = 30)

set.seed(1)
sub.obj <- runCluster(sub.obj, cluster.method = "som", xdim = 2, ydim = 3)
table(sub.obj@meta.data$cluster.id)

sub.obj <- runFastPCA(sub.obj)
sub.obj <- runTSNE(sub.obj)
sub.obj <- runDiffusionMap(sub.obj)
sub.obj <- runUMAP(sub.obj)
sub.obj <- buildTree(sub.obj, cluster.type = "som", dim.type = "tSNE")

plot(sub.obj@network$mst)

plot2D(sub.obj, item.use = c("CD34", "CD49f"), color.by = "stage", alpha = 1, main = "PCA", category = "categorical", show.cluser.id = T)

plotTree(sub.obj, color.by = "D8.percent", show.node.name = T, cex.size = 2) + scale_colour_gradientn(colors = c("#00599F", "#EEEEEE", "#FF3222"))

plotPseudotimeDensity(sub.obj, adjust = 1)

run.time <- read.xlsx("../dataset/RunningTime.xlsx")
p <- ggbarplot(run.time, x = "Function", y = "Time_5000", fill = "#0076D4",
               xlab = "", ylab = "Time (Seconds)", label = T) + theme_base()
p <- p + scale_y_continuous(limits = c(0,270), expand=c(0,0)) + coord_flip()
p <- p + theme(axis.text.x = element_text(angle = 90,hjust = 1,vjust = 1))
ggsave("../dataset/Time_5000.pdf", width = 6, height = 5)


###############
# merge FCS
fcsFiles <- paste0("../flowSpy-dataset/FCS/usecase2/D", c(0,2,4,6,8,10), ".fcs")

merged <- runExprsMerge(fcsFiles)



}


















